### PR TITLE
script: add description for the functionality of each opcode

### PR DESCRIPTION
**Pull request by ChrisCho-H**
While miniscript has a very kind explanation for it, original script does not.
I add a simple description for the functionality of each opcode.
It will help a lot for developers who want to do script programming.

### CODE REVIEW

**Comment by jamesob** [comment]
Concept ACK

Probably worth just squashing this down to 1 commit.

**Comment by Hyunhum** [comment]
Grateful for kind feedbacks, @sipa
I made commit to reflect feedbacks, and will squash this down to 1 commit if no more feedback to be approved!

**Review by jonatack — COMMENTED** [review]
Concept ACK

**Comment by RandyMcMillan** [comment]
Concept ACK

**Comment by kristapsk** [comment]
Concept ACK

**Comment by RandyMcMillan** [comment]
applying clang-formatting:
per `src/.clang-format`
may be useful

https://github.com/RandyMcMillan/bitcoin/commit/99e3dc536dca94dffd703e77572f6dca91f4bb26

File: src/script/script.h
Comment (john-moffett) [inline src/script/script.h:100]: We should probably be consistent in describing how Script deals with true/false, since it's later described as `if top stack value != 0`.  Maybe add a section along the lines of:

``` 
/** 
* Opcodes that take a true/false value will evaluate the following as false:
*     an empty vector
*     a vector (of any length) of all zero bytes 
*     a single byte of "\x80" ('negative zero')
*     a vector (of any length) of all zero bytes except the last byte is "\x80"
*
* Any other value will evaluate to true.
* /
```

I'd also say ` if the top stack value is true,`, and indicate that it consumes the stack item.

**Comment by Hyunhum** [comment]
clang-format applied and description for evaluating false added!

**Review by WM7586 — APPROVED** [review]

File: src/script/script.h
Comment (theStack) [inline src/script/script.h:85]:
```suggestion
    OP_PUSHDATA4 = 0x4e, // read the next 4 bytes as N and push the next N bytes as an array onto the stack
```

File: src/script/script.h
Comment (theStack) [inline src/script/script.h:158]: would also use the active form here
```suggestion
    OP_1ADD = 0x8b,      // add 1 to the top stack item
    OP_1SUB = 0x8c,      // subtract 1 from the top stack item
```

**Review by theStack — COMMENTED** [review]
Concept ACK

**Comment by darosior** [comment]
>  It will help a lot for developers who want to do script programming.

No strong opinion, but this purpose is already served by this [detailed wiki article](https://en.bitcoin.it/wiki/Script#Opcodes) presenting how each opcode works along with a bit of history. Using a "opcode, input, output, description" table it is even probably better at explaining what each opcode does than we can ever get in code comments.

**Comment by Hyunhum** [comment]
Detailed wiki does help dev! However, it's just a complementary material. 
If simple description on script.h, 
1. Bitcoin repo itself leads the standard description of opcodes, does not depend on external ones
2. Newly added opcode can be described here first
3. dev can get info easier and publicly discuss more to utilize script

These are my personal opinion, and any opinion is welcome!

**Comment by 1440000bytes** [comment]
> > It will help a lot for developers who want to do script programming.
> 
> No strong opinion, but this purpose is already served by this [detailed wiki article](https://en.bitcoin.it/wiki/Script#Opcodes) presenting how each opcode works along with a bit of history. Using a "opcode, input, output, description" table it is even probably better at explaining what each opcode does than we can ever get in code comments.

Agree with this. So many single line comments look weird. If wiki has some outdated or incorrect information, maybe that can be updated and use its link in the comment above all opcodes.

**Review by brunoerg — COMMENTED** [review]
Concept ACK

**Comment by RandyMcMillan** [comment]
NIT

```bash
cd src/ && clang-format --lines 68:148 -i script/script.h
```

```c
diff --git a/src/script/script.h b/src/script/script.h
index 35211ff376..5143593029 100644
--- a/src/script/script.h
+++ b/src/script/script.h
@@ -66,17 +66,16 @@ std::vector<unsigned char> ToByteVector(const T& in)
 }
 
 /** Script opcodes */
-enum opcodetype
-{
-/**
- * Opcodes that take a true/false value will evaluate the following as false:
- *     an empty vector
- *     a vector (of any length) of all zero bytes
- *     a single byte of "\x80" ('negative zero')
- *     a vector (of any length) of all zero bytes except the last byte is "\x80"
- *
- * Any other value will evaluate to true.
- */
+enum opcodetype {
+    /**
+     * Opcodes that take a true/false value will evaluate the following as false:
+     *     an empty vector
+     *     a vector (of any length) of all zero bytes
+     *     a single byte of "\x80" ('negative zero')
+     *     a vector (of any length) of all zero bytes except the last byte is "\x80"
+     *
+     * Any other value will evaluate to true.
+     */
     // push value
     OP_0 = 0x00,         // push "" onto the stack (which is an empty array of bytes)
     OP_FALSE = OP_0,     // push "" onto the stack (which is an empty array of bytes)


```

**Comment by Hyunhum** [comment]
> NIT
> 
> ```shell
> cd src/ && clang-format --lines 68:148 -i script/script.h
> ```
> 
> ```c
> diff --git a/src/script/script.h b/src/script/script.h
> index 35211ff376..5143593029 100644
> --- a/src/script/script.h
> +++ b/src/script/script.h
> @@ -66,17 +66,16 @@ std::vector<unsigned char> ToByteVector(const T& in)
>  }
>  
>  /** Script opcodes */
> -enum opcodetype
> -{
> -/**
> - * Opcodes that take a true/false value will evaluate the following as false:
> - *     an empty vector
> - *     a vector (of any length) of all zero bytes
> - *     a single byte of "\x80" ('negative zero')
> - *     a vector (of any length) of all zero bytes except the last byte is "\x80"
> - *
> - * Any other value will evaluate to true.
> - */
> +enum opcodetype {
> +    /**
> +     * Opcodes that take a true/false value will evaluate the following as false:
> +     *     an empty vector
> +     *     a vector (of any length) of all zero bytes
> +     *     a single byte of "\x80" ('negative zero')
> +     *     a vector (of any length) of all zero bytes except the last byte is "\x80"
> +     *
> +     * Any other value will evaluate to true.
> +     */
>      // push value
>      OP_0 = 0x00,         // push "" onto the stack (which is an empty array of bytes)
>      OP_FALSE = OP_0,     // push "" onto the stack (which is an empty array of bytes)
> ```

Because it's global comment for opcodes, I didn't indent. Reference might be [this](https://github.com/bitcoin/bitcoin/blob/4087c156535dd3e939baa37384347906dbb4d78b/src/script/script.h#L233).

**Comment by RandyMcMillan** [comment]
ACK 40f36d0a3dac568bc9b05dea021d83986d8add84

**Comment by ChrisCho-H** [comment]
Need more reviewers to be merged?

File: src/script/script.h
Comment (Sjors) [inline src/script/script.h:194]: I still have no idea how `OP_CODESEPARATOR` works, but that will take more than one code comment to explain.

**Review by Sjors — APPROVED** [review]
ACK 40f36d0a3dac568bc9b05dea021d83986d8add84

It matches my (limited) understanding of what these opcodes do. I also walked through `EvalScript()` to compare.

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/script/script.h
Comment (vostrnad) [reply src/script/script.h:197]: Additionally, at least according to Bitcoin Optech, "multisignature" should not be confused with "multisig": https://bitcoinops.org/en/topics/multisignature/

Comment (Sjors) [reply src/script/script.h:197]: I guess we can say "for multiple signatures" (but nobody reading this deep inside the codebase will be confused)

**Comment by ChrisCho-H** [comment]
Added followup commit for tapscript(+ suggestion from @vostrnad )!

**Comment by Sjors** [comment]
re-ACK 45926671dbf9e4919bf6aaee90e0d90f89830ce5

Nit: consider moving the typo fixes to the first commit (or just squash both).

**Review by Sjors — APPROVED** [review]
re-ACK 86b43d7fdd086c308f58414f3718a9cbf106bd05

**Comment by ChrisCho-H** [comment]
Looking for code reviewers! appreciate voluntary dedication in advance 🙏

File: src/script/script.h
Comment (theStack) [inline src/script/script.h:127]:
```suggestion
    OP_IFDUP = 0x73,        // if top stack value is true, duplicate it
```

File: src/script/script.h
Comment (theStack) [inline src/script/script.h:151]: Should probably also mention here that the inputs are popped from the stack? E.g. something like
`pop two top stack items and push 1 if they are equal, 0 otherwise`

**Comment by ChrisCho-H** [comment]
Thanks! I've fixed it.

**Comment by ChrisCho-H** [comment]
rebase done!

**Comment by ChrisCho-H** [comment]
@MarcoFalke would u mind to review this?

**Comment by ryanofsky** [comment]
This PR has two stale acks and seems like a safe documentation change. I think it would be good to merge if original reviewers want to reack or anyone else wants to review.

**Comment by ChrisCho-H** [comment]
Thx! I updated

**Review by vostrnad — COMMENTED** [review]
The more I look at this (and especially after the switch to notation), the more I think the descriptions should each be on a separate line above the opcode in a block comment (`/** ... */`). It would allow for longer descriptions where needed without ever exceeding the soft 100-per-line character limit.

The block comments could now include (where appropriate):
* Both a textual description and notation (frankly, I find the notation-only comments quite difficult to read).
* Historical context for an opcode (when was it enabled/disabled).
* Differences between tapscript and pre-tapscript.

Additionally, block comments above declarations are shown by many IDEs when hovering above a symbol, providing immediate access to documentation. Inline comments don't do this.

I've drafted a few examples for these longer descriptions:

```cpp
/**
 * Mark transaction as invalid if the top stack item is greater than
 * the transaction's locktime. The top stack item is not removed.
 * 
 * Previously OP_NOP2. Specified in BIP65 and deployed in December 2015
 * at block 388380 as a soft fork.
 */
OP_CHECKLOCKTIMEVERIFY = 0xb1,
/**
 * See OP_CHECKLOCKTIMEVERIFY.
 */
OP_NOP2 = OP_CHECKLOCKTIMEVERIFY,

/**
 * Mark transaction as invalid even when occurring in an unexecuted branch.
 * 
 * Previously concatenated two byte arrays. Disabled in August 2010.
 * 
 * Turned into OP_SUCCESS126 in tapscript.
 */
OP_CAT = 0x7e,

/**
 * Duplicate the top three stack items:
 * 
 * [ ... x0 x1 x2 ] -> [ ... x0 x1 x2 x0 x1 x2 ]
 */
OP_3DUP = 0x6f,
```

I understand this will make the section longer by a good few hundred lines, but if Bitcoin Core is to maintain in perpetuity a documentation for all opcodes, it should be at least of similar quality to the one at Bitcoin Wiki, and that is not achievable with short single-line descriptions.

**Comment by Sjors** [comment]
@vostrnad's comment makes sense to me. I'll hold off re-reviewing until that's either done or decided against.

File: src/script/script.h
Comment (RandyMcMillan) [inline src/script/script.h:169]: `pop two` doesn't quite work (IMO).

it makes the order somewhat ambiguous.

for example:

`pop pop add push` is more correct. no? 

Being more explicit in this way will make the `OP_SUB` make more sense, no?.

`pop pop subtract push`


for `OP_ADD` it isn't much of an issue.

`` 1 + 1 = 2 ``

`` 1 + -1 = 0 ``

`` -1 + 1 = 0 ``




for `OP_SUB` it may be an issue.

``1 - 1 = 0``

``1 - -1 = 0``

``-1 - 1 = -2``

*in common arithmetic notation

File: src/script/script.h
Comment (RandyMcMillan) [inline src/script/script.h:176]: see above.

better boolean notation here would be great.

**Comment by RandyMcMillan** [comment]
Concept ACK

**Comment by ChrisCho-H** [comment]
Seems no one's against @vostrnad suggestion, I convert all single lines to block comments, add textual description with notation and reference for OP_SUCCESS opcodes in case of tapscript.
I didn't include history much. In my opinion, description only for function itself looks clean and explicit which fits with main purpose of this.

**Comment by Sjors** [comment]
> Probably worth just squashing this down to 1 commit.

Indeed.

**Review by darosior — COMMENTED** [review]
Leaning toward Concept NACK. I'm sorry for the time you've invested in this (hope you got something out of it) but i think this is not a good use of review time and prone to bikeshedding.

**Comment by ChrisCho-H** [comment]
Understood. Thanks for everyone who put efforts into here tgh🙏

**Comment by ChrisCho-H** [comment]
For those who might look for this, I just created repo for it [here](https://github.com/ChrisCho-H/bitcoin-script).

**Comment by vostrnad** [comment]
For anyone intersted in documenting opcodes, there is now https://opcodeexplained.com/ ([GitHub repo here](https://github.com/thunderbiscuit/opcode-explained)).

**Pull request closed without merging**
